### PR TITLE
klippy: a controller that auto-instantiates packages

### PR DESCRIPTION
**Pull request by justinsb**
This controller will automatically create draft package instances where the bindings are satisfied.  If the bindings are sufficiently constraining, this can act as a guided workflow to help users do the next right thing.

### CODE REVIEW

File: porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+	// TODO: We should be able to cache this
+	blueprints, err := r.parseBlueprints(ctx, &blueprintPackageRevisions)
+	if err != nil {
+		return err
+	}
+
+	parentObjects, err := parseObjects(ctx, parentResources)
+	if err != nil {
+		return err
+	}
```
Comment (droot) [inline porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go:121]: parseObjects and parseBlueprints seems like functions could be useful in other places (porchclient++).

File: porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+
+	var allProposals []*api.PackageRevision
+	for _, blueprint := range blueprints {
+		proposals, err := r.buildProposals(ctx, parent, parentObjects, blueprint)
```
Comment (droot) [inline porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go:125]: seems like this is the part we would like to externalize (delegate to proposal makers).

File: porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+				Eval: &api.FunctionEvalTaskSpec{
+					Image:  image,
+					Config: bindingConfig,
+				},
+			})
+
+		}
+
+		proposals = append(proposals, proposal)
+	}
```
Comment (droot) [inline porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go:358]: lots of binding stuff that I skipped over. I am curious, where is `bind` function used that was taken as input ?

**Review by droot — COMMENTED** [review]
some drive by comments.

**Comment by johnbelamaric** [comment]
This is quite similar to what I am looking for with #3448. The idea is that a local-config resource will describe the "my package requires", and another one in other packages will describe the "my package provides". I want a dependency resolution controller that sees these in Draft package revisions and resolves them. Dependencies themselves are also surfaced as conditions, so that we can identify when all dependencies have been satisfied.

File: porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go (hunk 1 above)
Comment (justinsb) [reply porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go:121]: Yes - I feel like we should have them somewhere already.  I just don't know where...

File: porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go (hunk 3 above)
Comment (justinsb) [reply porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go:358]: Ooops - I forgot to do that - fixed in the latest update!

**Comment by justinsb** [comment]
> This is quite similar to what I am looking for with https://github.com/GoogleContainerTools/kpt/issues/3448. The idea is that a local-config resource will describe the "my package requires", and another one in other packages will describe the "my package provides". I want a dependency resolution controller that sees these in Draft package revisions and resolves them. Dependencies themselves are also surfaced as conditions, so that we can identify when all dependencies have been satisfied.

That's great news, thank @johnbelamaric .  I do expect we'll have multiple controllers making proposals.  I think it would be great if @yuwenma (or @natasha41575 ) tackled updating-the-container-image next.  I don't know whether your needs are close enough that we should extend this controller, or whether we should build a second controller - generally I think we'll evolve our understanding as we build out this functionality (better to have many specialized controllers, or fewer more generic ones?)

**Review by mortent — COMMENTED** [review]
Looks good to me, but can we add a README file that describes what it does and includes a simple example for testing it? I've found that pretty useful in the other controllers when I need to make changes. So something like this: https://github.com/GoogleContainerTools/kpt/tree/main/porch/controllers/workloadidentitybindings/config/samples

File: porch/controllers/klippy/pkg/controllers/klippy/rbac.go [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package klippy
+
+//+kubebuilder:rbac:groups=porch.kpt.dev,resources=packagerevisionresources,verbs=get;list;watch;create;update;patch;delete
```
[the remark this answers is not in the document]
Comment (justinsb) [reply porch/controllers/klippy/pkg/controllers/klippy/rbac.go:17]: Was mostly just trying it out to see how I/we felt about it.   The RBAC generated is the same either way.  I'll move it into the main controller, to follow the existing patterns.

File: porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+func (r *KlippyReconciler) parseBlueprints(ctx context.Context, packageRevisions *api.PackageRevisionList) ([]*blueprint, error) {
+	log := log.FromContext(ctx)
+
+	var blueprints []*blueprint
+
+	for i := range packageRevisions.Items {
+		packageRevision := &packageRevisions.Items[i]
+
+		if !strings.HasPrefix(packageRevision.Spec.PackageName, "infra/") {
+			log.Info("HACK: ignoring non-infra package", "package", packageRevision.Spec.PackageName)
```
[the remark this answers is not in the document]
Comment (justinsb) [reply porch/controllers/klippy/pkg/controllers/klippy/klippy_controller.go]: Good call, I removed the hack.  Instead we only apply this where packages actually have bindings, which achieves much the same thing (scoping just to the packages that are using bindings).

**Comment by justinsb** [comment]
Thanks @mortent - I added a README and was able to remove the hack.

**Review by mortent — APPROVED** [review]

**Pull request merged**